src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -907,6 +911,33 @@ def _infer_schema_for_file_format(
 
         return new_schema, schema_to_cast, read_file_transformations, None
 
+    def _infer_schema_from_user_input(self, user_schema: StructType, format: str):
+        if format != "JSON":
+            raise ValueError(
+                f"Currently only support user schema for JSON or XML format, got {format} instead"
+            )
+        schema_to_cast = []
+        transformations = []
+        new_schema = []
+        for field in user_schema.fields:
+            name = quote_name_without_upper_casing(field._name)
+            new_schema.append(
+                Attribute(
+                    name,
+                    field.datatype,
+                    field.nullable,
+                )
+            )
+            identifier = f"$1:{name}::{convert_sp_to_sf_type(field.datatype)}"
+            schema_to_cast.append((identifier, field._name))
+            transformations.append(sql_expr(identifier))
+        self._user_schema = StructType._from_attributes(new_schema)
+        # If the user sets transformations, we should not override this
+        self._infer_schema_transformations = transformations
+        self._infer_schema_target_columns = self._user_schema.names
```

Comment:
I think in `_infer_schema_for_file_format` we have the `try-except` logic, does it apply here?

Reply:
no, because there is no communication between server and client in this function